kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/storage/cloud/KieServerStateCloudRepository.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+package org.kie.server.services.impl.storage.cloud;
+
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.validation.constraints.NotNull;
+
+import com.thoughtworks.xstream.XStream;
+import com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider;
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import org.kie.server.api.KieServerConstants;
+import org.kie.server.api.model.KieContainerResource;
+import org.kie.server.api.model.KieServerConfigItem;
+import org.kie.server.services.impl.storage.KieServerState;
+import org.kie.server.services.impl.storage.KieServerStateRepository;
+import org.kie.soup.commons.xstream.XStreamUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KieServerStateCloudRepository implements KieServerStateRepository, CloudClientFactory {
+
+    public static final String ROLLOUT_REQUIRED = "services.server.kie.org/openshift-startup-strategy.rolloutRequired";
+    public static final String KIE_SERVER_STATE_UPDATED = "org.kie.server.state.updated";
```

Comment:
Where is this property configured?
I can see its usage just in test and in KieServerStateCloudRepository to check if any change was made, but I don't see where it is set.

Reply:
That 'KIE_SERVER_STATE_UPDATED' property has been removed as a result of the latest enhancement. It was designed as an explicit trigger for KieServer DC rollout. Right now, it is automatically dealt by KieContainerCommandService.  In response to certain commands, such as      CreateContainerCommand, DisposeContainerCommand, UpdateScannerCommand, and UpdateReleaseIdCommand, which will lead to a KieServer state change and a call to KieServerStateRepository.store() method, the KieServerStateCloudRepository will annotate the correspondent CM so as to trigger a DC rollout accordingly.

Reply:
Ok, another question then, why is the rollout limited just to the mentioned commands? Why not to rollout in every case when KieServerStateRepository.store() is called? (Same question as added below)

Reply:
Once KieServerStateRepository.store() being called, triggering a DC rollout is depending on whether the KieServer instance has completed initialization or not. For some reason, there are repetitive calls to store() method during the initialization phase that should NOT trigger any DC rollout, otherwise it will end up with infinite loop. One option is to intercept certain Command execution. Alternatively, as you suggested, we can leverage the isKieServerReady flag which is what has been implemented in the lastest commit.